From 85d9dfb83d49499b21fb3c97e4486bc944a2651e Mon Sep 17 00:00:00 2001
From: Eric Garver <eric@garver.life>
Date: Tue, 14 Oct 2025 15:59:05 -0400
Subject: [PATCH 40/48] v2.4.0: fix(nftables): ipset: add entries from GLib
 loop when idle

Sets with a large amount of entries can take a significant time to
apply. Use the GLib mainloop to add them in chunks when the loop is
idle. This allows dbus calls in between the chunks as the dbus
events/calls have higher priority.

Fixes: #1277
(cherry picked from commit 3874bafc427139e647829d7662577567343aceb6)
---
 src/firewall/core/nftables.py | 23 +++++++++++++++++++----
 1 file changed, 19 insertions(+), 4 deletions(-)

diff --git a/src/firewall/core/nftables.py b/src/firewall/core/nftables.py
index 8115bcb9d7f4..254fe7cfbebe 100644
--- a/src/firewall/core/nftables.py
+++ b/src/firewall/core/nftables.py
@@ -18,6 +18,9 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #
+
+from gi.repository import GLib
+
 import copy
 import json
 import ipaddress
@@ -1958,15 +1961,27 @@ class nftables(object):
         rules = []
         rules.extend(self.build_set_create_rules(set_name, type_name, create_options))
         rules.extend(self.build_set_flush_rules(set_name))
+        self.set_rules(rules, self._fw.get_log_denied())
+
+        def _idle_set_add_entries(rules):
+            try:
+                self.set_rules(rules, self._fw.get_log_denied())
+            except Exception as e:
+                log.error("While restoring ipset entries the following Error occurred:")
+                log.error(e)
 
-        # avoid large memory usage by chunking the entries
+        # Avoid large memory usage by chunking the entries. Additionally, add
+        # the entries from the GLib main loop when it's idle. This avoids
+        # blocking the main loop for too long.
+        #
         chunk = 0
+        rules = []
         for entry in entries:
             rules.extend(self.build_set_add_rules(set_name, entry))
             chunk += 1
             if chunk >= 1000:
-                self.set_rules(rules, self._fw.get_log_denied())
-                rules.clear()
+                GLib.idle_add(lambda x: _idle_set_add_entries(x), rules)
+                rules = []
                 chunk = 0
         else:
-            self.set_rules(rules, self._fw.get_log_denied())
+            GLib.idle_add(lambda x: _idle_set_add_entries(x), rules)
-- 
2.47.3

